Backport of CSP-related changes

jasondlee opened the pull request:
Backporting CSP-related changes from 5.x to the 4.0 branch.

jasondlee committed:
Backport of CSP-related changes

Backport from several Mojarra 5 commits
Backport Faces 5 TCK test of CSP for coverage in 4.0

jasondlee commented:
@arjantijms @BalusC Any thoughts on this?

BalusC commented on impl/src/main/java/com/sun/faces/renderkit/RenderKitUtils.java:
```diff
@@ -392,14 +395,16 @@ public static void renderSelectOnclick(FacesContext context, UIComponent compone
     }
 
-    // Renders the onclick handler for command buttons. Handles
+    // Renders the onclick event listent for command buttons. Handles
```
*listener

jasondlee commented:
@fcarriedos Thanks. I'll give that a look. My backport is, fwiw, a backport of all of the CSP-related changes from 5.x. I don't pretend to understand all of the changes, but I'll try to fix that. :)

jasondlee commented:
> > I fixed it in 5.0 #5631
> > @jasondlee you can now backport thas as well into your branch
> 
> Thanks! On it...

@BalusC Should this be working in the master branch? I built that branch and integrated that impl jar (and the 5.0 API jar) into WildFly locally so I could verify behavior and correctness, but, using @softwaresicario's reproducer above, it appears that commandLinks are broken if the CSP policy is applied.

```
Navigated to http://localhost:8080/reproducer-04209988/mytest.jsf
mytest.jsf:7 Executing inline script violates the following Content Security Policy directive 
'default-src 'self''.  Either the 'unsafe-inline' keyword, a hash 
('sha256-gRzq6dUwQTPnqxVhNkutQ5tMemn1eFOfoM1R3fr05Sc='), 
or a nonce ('nonce-...') is required to enable inline execution. Note also that 'script-src' 
was not explicitly set, so 'default-src' is used as a fallback. The action has been blocked.
```

BalusC commented:
And this https://github.com/eclipse-ee4j/mojarra/pull/5653/files also needs to be backported. That'll be the final part I think :)

BalusC commented:
> but it's failing in the IT

Which IT exactly is failing with which error exactly?

jasondlee commented:
> Aah .. you got 0xFFFD, yeah ok I've adjusted the IT [2890b25](https://github.com/eclipse-ee4j/mojarra/commit/2890b25c0c1eb0760a96281203065f7499ddc3a7)

I'm always hesitant to change a test, especially an old one. :) That said, I'm still getting an error:

```
[ERROR] Tests run: 2, Failures: 1, Errors: 0, Skipped: 0, Time elapsed: 16.58 s <<< FAILURE! -- in org.eclipse.mojarra.test.issue5464.Issue5464IT
[ERROR] org.eclipse.mojarra.test.issue5464.Issue5464IT.testFFFE -- Time elapsed: 0.213 s <<< FAILURE!
org.opentest4j.AssertionFailedError: Output should not contain U+FFFE: Result: f￾oo ==> expected: <false> but was: <true>
	at org.junit.jupiter.api.AssertionFailureBuilder.build(AssertionFailureBuilder.java:151)
	at org.junit.jupiter.api.AssertionFailureBuilder.buildAndThrow(AssertionFailureBuilder.java:132)
	at org.junit.jupiter.api.AssertFalse.failNotFalse(AssertFalse.java:63)
	at org.junit.jupiter.api.AssertFalse.assertFalse(AssertFalse.java:36)
	at org.junit.jupiter.api.Assertions.assertFalse(Assertions.java:239)
	at org.eclipse.mojarra.test.issue5464.Issue5464IT.testFFFE(Issue5464IT.java:45)
```

I'm also not entirely sure how the changes in the PR would affect unicode handling, though, perhaps, the code path has changed for the xhr requests?

jasondlee commented:
@BalusC @softwaresicario I have backported, I think, all of the upstream changes as well as porting the Spec1950Test from TCK to the 4.0 integration test suite to get coverage for 4.0 since it won't be covered in the TCK.

I *did* use Claude to work through some JS-related test failures, and while I have pushed those to GH, I'm still reviewing those, but I wanted to see how these changes fare on a machine other than my own.

More scrutiny of all of these changes are 100% welcome. :)

BalusC commented:
Backport seems to miss the following:
- https://github.com/eclipse-ee4j/mojarra/pull/5672
- https://github.com/eclipse-ee4j/mojarra/pull/5678
- https://github.com/eclipse-ee4j/mojarra/pull/5696

BalusC committed:
Have Claude sync to Faces API 5.0 and Mojarra master branches and catch
up inconsistencies in the backport

BalusC commented:
There were only a few inconsistencies which I ironed out: https://github.com/eclipse-ee4j/mojarra/pull/5606/commits/d890b5a24ec64f94ce271cab465c437951626aaf

However, on second thought, I cannot accept the changes to API classes as that contaminates the 4.0 (and 4.1) API. It would basically falsely claim that these exist in 4.0 spec. These constants/methods have to remain impl-specific. Only when merging the changes into 5.0 branch then we can remove the impl-specific changes and rely on real spec/api based impl. In other words, ResourceHandler and ResourceHandlerWrapper have to be reverted, the constants/methods moved into ResourceHandlerImpl and the context parameters to be com.sun.faces prefixed instead of jakarta.faces. I will take care of this with help of Claude.

BalusC committed:
Sync Spec1590IT as well

BalusC committed:
Revert API changes and keep CSP impl-specific (with help of Claude)

BalusC committed:
Add missing backports from 4241c65 and 082789b with help of Claude;
This change ensures that generated script elements are not rendered
nested inside the element but only after the element as nested script
elements violate html spec in case of <a> and <input> elements

BalusC commented:
- All backports are in place.
- API changes have been reverted so it's impl specific only (will be "restored" in 5.0).
- context params are `com.sun.faces.enableCspNonce` and `com.sun.faces.cspPolicy` (will be "restored" in 5.0).
- IT test passes.

Ready for merge :)

jasondlee commented:
> There were only a few inconsistencies which I ironed out: [d890b5a](https://github.com/eclipse-ee4j/mojarra/commit/d890b5a24ec64f94ce271cab465c437951626aaf)
> 
> However, on second thought, I cannot accept the changes to API classes as that contaminates the 4.0 (and 4.1) API. It would basically falsely claim that these exist in 4.0 spec. These constants/methods have to remain impl-specific. Only when merging the changes into 5.0 branch then we can remove the impl-specific changes and rely on real spec/api based impl. In other words, ResourceHandler and ResourceHandlerWrapper have to be reverted, the constants/methods moved into ResourceHandlerImpl and the context parameters to be com.sun.faces prefixed instead of jakarta.faces. I will take care of this with help of Claude.

That's a good catch. Thanks for the clean up and the merge. I'll use this PR to create a backport for 4.1.

BalusC commented:
@jasondlee it's already merged in 4.1: https://github.com/eclipse-ee4j/mojarra/commit/54720bd4ceafaacf56fc262becd18b196720be6b

jasondlee commented:
> @jasondlee it's already merged in 4.1: [54720bd](https://github.com/eclipse-ee4j/mojarra/commit/54720bd4ceafaacf56fc262becd18b196720be6b)

Even better. Thanks for handling that, and all the help with this PR. :) If there are no objections, I'm going to release an update for both 4.x versions.

BalusC commented:
TCK job in Eclipse CI is not working since changes in publishing flow due to migration of Sonatype to Maven Central (cc: @arjantijms ) so you need to manually run TCK beforehand for 4.0 as well as 4.1. If TCK passes for you then you only need to execute "build and stage" job (autoPublish flag is currently true so no additional action is needed).

jasondlee commented:
Does this process still apply? https://github.com/eclipse-ee4j/mojarra/blob/master/RELEASE.md

BalusC commented:
Nope. Only job 1 at Jenkins works. Job 1 currently basically already publishes (because of recently added autoPublish flag in pom.xml). Jobs 2 and 3 don't anymore work (they still rely on Sonatype) and should therefore be skipped. Full TCK should be executed manually at your machine (in confidence). All this is because of the migration from Sonatype Nexus to Maven Central about 8 months ago.

From the RELEASE.md this thus means you can skip the following steps:
- 1
- 7-10
- 13-18

We still need to catch up the build scripts and RELEASE.md. We never had real time for this. I'll try to delegate this to Claude this week.

jasondlee commented:
How does this draft look? https://github.com/eclipse-ee4j/mojarra/releases/edit/untagged-c5885c388e36ba86436a Not sure I've done that part before :)

BalusC commented:
LGTM :)

BalusC commented:
I'm currently testing 4.1.8 release with the new JenkinsFile: https://ci.eclipse.org/mojarra/job/mojarra-release/1/console
